### PR TITLE
Integrate alchemy

### PR DESCRIPTION
**Pull request by johnmosesman**
Uses alchemy to get metadata for this scan-an-nft component. On opensea paste fetches the NFT image. Analyze just goes to blank thing atm.

image

### CODE REVIEW

File: src/app.html [hunk 1]
```diff
@@ -10,6 +10,9 @@
 		<link rel="preconnect" href="https://fonts.googleapis.com">
 		<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
 		<link href="https://fonts.googleapis.com/css2?family=Fira+Sans+Extra+Condensed:wght@800&family=Fira+Sans:wght@300&display=swap" rel="stylesheet">
+
+		<!-- TODO: load this via package -->
+		<script src="https://cdn.jsdelivr.net/npm/@alch/alchemy-web3@latest/dist/alchemyWeb3.min.js"></script>
```
Comment (ryanlabouve) [inline src/app.html:15]: What issue we running into loading this via normal routes? This the node specific function issue?

File: src/lib/components/ScanAnNft.svelte [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  import { onMount } from 'svelte';
+  import env from '$lib/constants/env';
+
+  let hasEvaluatedNft: boolean = false;
+  let openseaUrl: string = '';
+  let imageUrl: string = '';
+  let contractAddress: string = '';
+  let tokenId: string = '';
+
+  function parseAndSetFromOpenseaUrl(url) {
```
Comment (ryanlabouve) [inline src/lib/components/ScanAnNft.svelte]: Can you add received and returned types here?

File: src/lib/components/ScanAnNft.svelte [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    parseValuesFromOpenseaUrl(url);
+    fetchImage(contractAddress, tokenId);
+  }
+
+  function parseValuesFromOpenseaUrl(url: string) {
+		// https://opensea.io/assets/:nftContractAddress/:tokenId
+		let openseaAddressAndIdRe = /https:\/\/opensea.io\/assets\/(\w*)\/(\w*)/g;
```
Comment (ryanlabouve) [inline src/lib/components/ScanAnNft.svelte:18]: Hello old friend https://github.com/startupdotdev/forever721-ext/blob/main/src/lib/utils/open-sea.ts#L13

File: src/lib/components/ScanAnNft.svelte [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+			contractAddress = address;
+			tokenId = id;
+		}
+  }
+
+  async function fetchImage(_contractAddress: string, _tokenId: string): Promise<void> {
+    console.log("fetchImage()");
+
+    try {
+      const web3Alchemy = AlchemyWeb3.createAlchemyWeb3(env.alchemyApiKey);
```
Comment (ryanlabouve) [inline src/lib/components/ScanAnNft.svelte:34]: This will expose the env var to the client

File: src/lib/components/ScanAnNft.svelte [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      const result = await web3Alchemy.alchemy.getNftMetadata({
+        contractAddress: _contractAddress,
+        tokenId: _tokenId
+      });
+
+      console.log("fetchImage result:", result);
+
+      if (result && result.metadata && result.metadata.image) {
```
Comment (ryanlabouve) [inline src/lib/components/ScanAnNft.svelte]: We have safety operator avail `result?.metadata?.image`

File: src/lib/components/ScanAnNft.svelte [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+		if (address && id) {
+			openseaUrl = url;
+			contractAddress = address;
+			tokenId = id;
+		}
+  }
+
+  async function fetchImage(_contractAddress: string, _tokenId: string): Promise<void> {
+    console.log("fetchImage()");
```
Comment (ryanlabouve) [inline src/lib/components/ScanAnNft.svelte:29]: You may want to reset the variables you mutate in the function upon invocation. Could optionally put it in a loading state.

File: src/lib/components/ScanAnNft.svelte [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+      <svg xmlns="http://www.w3.org/2000/svg" class="text-white h-16 w-16" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M19 11H5m14 0a2 2 0 012 2v6a2 2 0 01-2 2H5a2 2 0 01-2-2v-6a2 2 0 012-2m14 0V9a2 2 0 00-2-2M5 11V9a2 2 0 012-2m0 0V5a2 2 0 012-2h6a2 2 0 012 2v2M7 7h10"></path></svg>
+    </div>
+  {/if}
+
+  <form on:submit|preventDefault={() => evaluateNft('hi', 'hi') }>
+    <div class='mb-4'>
+      <p class='mb-1'>OpenSea URL</p>
+      <input
+        class='border border-gray-400 rounded px-2 py-1 w-full'
+        on:paste={(e) => parseAndSetFromOpenseaUrl(e.clipboardData.getData('text'))}
```
Comment (ryanlabouve) [inline src/lib/components/ScanAnNft.svelte:116]: This looks familiar ;)

File: src/lib/constants/env.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export default {
+  alchemyApiKey: import.meta.env.VITE_ALCHEMY_API_URL
```
Comment (ryanlabouve) [inline src/lib/constants/env.ts:2]: Anything that begins with `import.meta.env.VITE_*` is avail to client.

File: src/app.html (hunk 1 above)
Comment (johnmosesman) [reply src/app.html:15]: Yeah I messed with it for a bit but didn't get installing via package to work yet

File: src/lib/components/ScanAnNft.svelte (hunk 3 above)
Comment (johnmosesman) [reply src/lib/components/ScanAnNft.svelte:18]: new open source package? 😄

File: src/lib/components/ScanAnNft.svelte (hunk 4 above)
Comment (johnmosesman) [reply src/lib/components/ScanAnNft.svelte:34]: good point, will wait to deploy until the api is in place

**Comment by johnmosesman** [comment]
Will do some function cleanup/reorg in a different PR

**Pull request merged**